akka-typed/src/main/scala/akka/typed/ScalaProcess.scala

```
@@ -0,0 +1,538 @@
+
+    def loopInf[S]: NextLoopInf[S] = nextLoopInf.asInstanceOf[NextLoopInf[S]]
+    trait NextLoopInf[S] {
+      def apply[U](body: OpDSL { type Self = S } ⇒ Operation[S, U]): Operation[S, Nothing] = {
+        lazy val l: Operation[S, Nothing] = unit[S, OpDSL { type Self = S }].flatMap(body).flatMap(_ ⇒ l)
+        l
+      }
+    }
+    private object nextLoopInf extends NextLoopInf[Nothing]
+
+    def loop[S]: NextLoop[S] = nextLoop.asInstanceOf[NextLoop[S]]
+    trait NextLoop[S] {
+      def apply[Out](n: Int)(body: OpDSL { type Self = S } ⇒ Operation[S, Out]): Operation[S, List[Out]] = {
+        require(n > 0, "number of iterations must be positive")
+        def step(n: Int, acc: List[Out]): Operation[S, List[Out]] =
+          unit[S, OpDSL { type Self = S }]
+            .flatMap(body)
+            .flatMap {
+              case result if n == 1 ⇒ Return((result :: acc).reverse)
+              case result           ⇒ step(n - 1, result :: acc)
+            }
+        step(n, Nil)
+      }
+    }
+    private object nextLoop extends NextLoop[Nothing]
+
+    def apply[T]: Next[T] = next.asInstanceOf[Next[T]]
+
+    trait Next[T] {
+      def apply[U](body: OpDSL { type Self = T } ⇒ Operation[T, U]): Operation[T, U] =
+        unit[T, OpDSL { type Self = T }].flatMap(body)
+    }
+    private object next extends Next[Nothing]
+
+    trait NextStep[T] {
+      def apply[U](mailboxCapacity: Int, body: OpDSL { type Self = T } ⇒ Operation[T, U])(implicit opDSL: OpDSL): Operation[opDSL.Self, U] =
+        Call(Process("nextStep", Duration.Inf, mailboxCapacity, body(null)))
+    }
+    private[typed] object nextStep extends NextStep[Nothing]
+  }
+
+  /*
+   * Terminology:
+   *
+   *  - a Process has a 1:1 relationship with an ActorRef
+   *  - an Operation is a step that a Process takes and that produces a value
+   *  - Processes are concurrent, but not distributed: all failures stop the entire Actor
```

Yeah, exactly - I meant the "supervision combinators" can be used to get a resume rather easily etc. I agree on the fail fast thing too. If I have "steps" then i can model them as steps of different behaviours I guess? Will get to that part soon in the PR I guess.

(Example being: command(fetch site A, fetch site B), so I want to store the results individually)